[chunzkok] iP

chunzkok opened the pull request:
# Clippy
## Why Clippy, you may ask?
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))
- It's ~lightweight~ *tiny*
- It's easy to learn
- It's useful to track your tasks!

## So how can you get started?
1. Download the .jar file [here](https://github.com/chunzkok/ip/releases/)
2. Double click on the downloaded file
3. Add your tasks using one of the three commands: `todo <description>`, `deadline <description> <date to complete by>` or `event <description> <start date> <end date>`.
4. Remove any tasks by typing in `delete <task ID>`
5. How to check your task and their task IDs? Use this command: `list`
6. When you're done, just end off with a `bye` and Clippy will save your data for your next visit 😎 

## Is this suitable for me?
Sure! If you have the following traits, you are **definitely** going to like this app:
- [ ] prefers typing over clicking
- [ ] has difficulties keeping track of tasks to complete
- [ ] always has a computer at hand

## What is Clippy developed using?
Java! Here is a small code snippet of how Clippy is initialised.
```Java
    public Clippy() {
        this.ui = new Ui(">>>", "###");
        this.storage = new Storage(this.ui);
        this.taskList = new TaskList(storage.loadState());
    }
```

Commit:
Add Gradle support

chunzkok committed:
Change bot logo and add .class files into .gitignore

chunzkok committed:
Add private method parseCommand to handle user commands

chunzkok committed:
Add prettyPrint method to print bot output with standardised format

chunzkok committed:
Add main driver code - accept input until 'bye' seen

chunzkok committed:
Implement 'list' command

chunzkok committed:
Implement [add] functionality

chunzkok committed:
Create Task class

chunzkok committed:
Modify Clippy to use Task in its toDoList

chunzkok committed:
Add 'mark' and 'unmark' commands

chunzkok committed:
Add ToDo class

chunzkok committed:
Add Deadline class

chunzkok committed:
Add Event class

chunzkok committed:
Implement 'todo' command

chunzkok committed:
Implement 'deadline' command

chunzkok committed:
Implement 'event' command

chunzkok committed:
Update text UI testing

chunzkok committed:
Added 3 custom exceptions regarding invalid user input

chunzkok committed:
Add 2 more exceptions regarding 'deadline' and 'event' comands

chunzkok committed:
Fix newly added Exceptions

chunzkok committed:
Add ClippyException handling

chunzkok committed:
Add checks that raise exceptions upon encountering invalid input

chunzkok committed:
Relocate ClippyException catching to let program continue on ClippyExceptions

chunzkok committed:
Added invalid command checking

chunzkok committed:
Add 'delete' command

kchunzhi committed:
Add methods to get CSV string from Deadlines, Events and Tasks + method to parse CSV lines

kchunzhi committed:
Add functionality for program to load saved state

kchunzhi committed:
Add state-saving functionality after 'bye' command + fix incorrect state saving method

yitong241 commented on src/main/java/ClippyInvalidEventException.java:
```diff
@@ -0,0 +1,6 @@
+        super("cannot recognise event start and end time - please specify it as such: \n" +
+                "event <description> /from <start> /to <end>");
+    }
+}
```
Seperation of different kinds of exception is good 👍

chunzkok committed:
Add date recognition support for Deadlines

chunzkok committed:
Add date recognition support for Events

chunzkok committed:
Add new print format to denote system messages

chunzkok committed:
Merge branch 'branch-Level-7'

chunzkok committed:
Merge branch 'branch-Level-8'

yitong241 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,123 @@
+     * [add] is invoked whenever any string that does not match other commands is entered.
+     * @param   command a string containing the command entered by the user
+     * @return          true if programme should continue accepting further commands, else false
+     */
```
It is good to have detailed java doc like this👍

yitong241 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public String toString() {
+        // only show day of the week + day of the month + month
+        return "[D]" + super.toString() +
```
It is good to break the line when it is too long 👍

yitong241 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,216 @@
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.io.File;
```
It is good to import one by one instead of using wildcard 👍

yitong241 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,216 @@
+                        int byIndex = command.indexOf("/by ");
+                        if (byIndex == -1 || command.length() < byIndex + 4) {
+                            throw new ClippyMissingDeadlineException();
+                        }
```
K&R style brackets is well used, good job!

yitong241 reviewed (commented):
The code and javadoc are easy to read and understand. Well done!

chunzkok committed:
Merge branch 'add-gradle-support' of https://github.com/chunzkok/ip

chunzkok committed:
Add Gradle support

chunzkok committed:
Add Ui class abstraction

chunzkok committed:
Add CommandType Enum

chunzkok committed:
Add Command abstract class

chunzkok committed:
Separate out Storage class

chunzkok committed:
Separate out TaskList class

chunzkok committed:
Separate out 'bye' command handler

chunzkok committed:
Separate out 'mark' command handler

chunzkok committed:
Separate out 'unmark' command handler

chunzkok committed:
Separate out 'todo', 'deadline' and 'event' command handlers

chunzkok committed:
Separate out 'list' command handler

chunzkok committed:
Separate out 'delete' command handler

chunzkok committed:
Modify command handlers' code for more readable code in the future

chunzkok committed:
Separate out Parser class

chunzkok committed:
Integrate Ui, Storage, Parser and TaskList classes into Clippy

chunzkok committed:
Change Duke class to Clippy

chunzkok committed:
Refactor java files into packages

chunzkok committed:
Add JUnit Integration + Tests for Deadline, Event, ToDo

chunzkok committed:
Add JavaDoc for command handlers

chunzkok committed:
Add JavaDoc for all files

chunzkok committed:
Add ind command functionality

chunzkok committed:
Merge branch 'branch-A-JavaDoc'

chunzkok committed:
Merge branch 'branch-Level-9'

chen-jerry-junior commented on src/main/java/clippy/task/Task.java:
```diff
@@ -0,0 +1,100 @@
+        Task result = null;
+        try {
+            switch (arguments[0]) {
+                case "D":
```
I think there shouldn't be more indentations for "case" under "switch"

chen-jerry-junior commented on src/main/java/clippy/task/TaskList.java:
```diff
@@ -0,0 +1,122 @@
+     * @param query Keyword(s) to be used to filter the Tasks.
+     * @return A list of tasks that contain the search query in their descriptions.
+     */
+    public List<Task> find(String query) {
```
I like your idea to utilize the stream().filter() method here

chen-jerry-junior commented on src/main/java/clippy/command/Parser.java:
```diff
@@ -0,0 +1,120 @@
+     * @param dates Dates, if any, to be used in the command.
+     * @return
+     */
+    private static Command dispatch(CommandType commandType, String[] args, LocalDate... dates) {
```
I like your idea of creating a dispatch() method to cooperate with your parse method, but will it become a better encapsulation if Parser class is not under the packet command?

chen-jerry-junior commented on src/main/java/clippy/task/Task.java:
```diff
@@ -0,0 +1,100 @@
+ */
+public abstract class Task {
+    private String description;
+    private boolean done;
```
will it be better if the naming becomes "isDone"?

chen-jerry-junior reviewed (commented):
LGTM, almost. Just a few nits to fix.

kchunzhi committed:
Add CheckStyle support + fix checkstyle errors

kchunzhi committed:
Add JavaFx integration and setup basic GUI layout

kchunzhi committed:
Add proper layout and event handlers to capture user input

kchunzhi committed:
Add custom DialogBox with factory methods for user/Clippy dialogue boxes

kchunzhi committed:
Refactor JavaFX code using FXML

kchunzhi committed:
Fix Checkstyle errors

chunzkok committed:
Link GUI to Clippy for proper responses

chunzkok committed:
Fix Checkstyle issues

chunzkok committed:
Add assert statements to document important assumptions

chunzkok committed:
Storage class: extract multiple nested code chunks

The storage class has many nested code chunks, especially in the
loadState() and saveState() methods,  leading to
arrowhead-style code and poor SLAP.

This is undesirable as it makes it harder to read and understand what is
going on in the Storage class.

Extracting the nested code into smaller helper methods allows us to give
the code chunks 'names' that can help in describing what it does. This
also makes the code more self-explanatory, reducing the need for
redundant comments.

Let's pull out all the code segments relating to writing data into the
save file into a method Storage::writeData.

Let's pull out the handling of missing save directories and save files
into the methods Storage::handleMissingSaveDir and
Storage::createNewSaveFile.

Let's extract out the functionality to load an existing save file into
the method Storage::loadExistingSaveFile.

Extracting out these code chunks into methods with self-explanatory
names helps to improve code readability and improve ease of debugging.

chunzkok committed:
Merge pull request #3 from chunzkok/branch-A-CodeQuality

Storage class: extract multiple nested code chunks

chunzkok committed:
Merge branch 'master' into branch-A-Assertions

chunzkok committed:
Merge pull request #2 from chunzkok/branch-A-Assertions

Add assert statements to document important assumptions

chunzkok committed:
GitHub Actions: Set up CI for project

On every push/merge, there is a chance of regression when the code
changes.

This could be catastrophic if left undetected for multiple commits which
might even build on the broken functionality.

Let's set up a simple continuous integration script that runs on GitHub
Actions to ensure that such regressions are caught and rectified as
early as possible, if any.

This is because GitHub actions allow us to do automated tesing on every
push/merge, which helps as manual testing is tedious and may not be as
consistent as automated tests.

chunzkok committed:
Fix checkstyle errors and remove CI text output checks

chunzkok committed:
Fix Checkstyle errors for Test scripts

chunzkok committed:
Add task priorities with default priority NONE

chunzkok committed:
Add new 'priority' command handler

chunzkok committed:
Fix Checkstyle issues

chunzkok committed:
Add title to displayed window

chunzkok committed:
Reformat command output to accomodate for new GUI

chunzkok committed:
Add user guide for Clippy

chunzkok committed:
Update Unit Tests for Tasks to reflect new Priority state

chunzkok committed:
Fix bug where program crashes for first-time users
